File: objc-demo/ChatDemo-UI3.0/Code/Model/EMConversationModel.m
/************************************************************
 *  * Hyphenate
 * __________________
 * Copyright (C) 2016 Hyphenate Inc. All rights reserved.
 *
 * NOTICE: All information contained herein is, and remains
 * the property of Hyphenate Inc.
 */

#import "EMConversationModel.h"

#import "EMUserProfileManager.h"

@implementation EMConversationModel

- (instancetype)initWithConversation:(EMConversation*)conversation
{
    self = [super init];
    if (self) {
        _conversation = conversation;
        
        NSString *subject = [conversation.ext objectForKey:@"subject"];
        if ([subject length] > 0) {
            _title = subject;
        }
        
        if (_conversation.type == EMConversationTypeGroupChat) {
            NSArray *groups = [[EMClient sharedClient].groupManager getJoinedGroups];
            for (EMGroup *group in groups) {
                if ([_conversation.conversationId isEqualToString:group.groupId]) {
                    _title = group.subject;
                    break;
                }
            }
        }
        
        if ([_title length] == 0) {
            _title = _conversation.conversationId;
        }
    }
    return self;
}

- (NSString*)title
{
    if (_conversation.type == EMConversationTypeChat) {
        return [[EMUserProfileManager sharedInstance] getNickNameWithUsername:_conversation.conversationId];
    } else {
        return _title;
    }
}

@end

File: swift-demo/Hyphenate-Demo-Swift/Code/Models/EMConversationModel.swift
//
//  EMConversationModel.swift
//  Hyphenate-Demo-Swift
//
//  Created by dujiepeng on 2017/6/14.
//  Copyright © 2017 dujiepeng. All rights reserved.
//

import UIKit
import Hyphenate

class EMConversationModel: NSObject, EMRealtimeSearchUtilDelegate {
    
    var _title:String?
    var conversation:EMConversation?
    
    init(conversation con: EMConversation) {
        conversation = con
        let subject = con.ext?["subject"] as? String
        if subject != nil {
            _title = subject
        }
        
        if conversation?.type == EMConversationTypeGroupChat {
            let groups = EMClient.shared().groupManager.getJoinedGroups()
            for group in groups! {
                if conversation?.conversationId == (group as! EMGroup).groupId {
                    _title = (group as! EMGroup).subject
                    break
                }
            }
        }
        
        if _title?.characters.count == 0 {
            _title = conversation?.conversationId
        }
    }
    
    func searchKey() -> String? {
        return title()
    }
    
    func title() -> String? {
        if conversation!.type == EMConversationTypeChat  {
            return EMUserProfileManager.sharedInstance.getNickNameWithUsername(username: conversation!.conversationId)
        } else {
            return _title
        }
    }
    
    func isTop()->Bool {
        if conversation?.ext?["isTop"] != nil {
            let numIsTop = conversation?.ext?["isTop"] as! NSNumber
            return numIsTop.boolValue ? true : false
        }
        
        return false
    }
    
    func setIsTop(isTop: Bool) {
        var dic = conversation?.ext
        if dic == nil {
            dic = Dictionary()
        }
        dic!["isTop"] = NSNumber.init(value: isTop)
        conversation?.ext = dic
    }
    
    func removeComplation (completion aCompletionBlock: @escaping () -> Void){
        EMClient.shared().chatManager.deleteConversation(conversation?.conversationId, isDeleteMessages: true) { (aConversationId, aError) in
           aCompletionBlock()
        }
    }
}
